[libc++][math] Mathematical Special Functions: Hermite Polynomial

PaulXiCao opened the pull request:
Implementing the Hermite polynomials which are part of C++17's mathematical special functions. The goal is to get early feedback which will make implementing the other functions easier. Integration of functions in chunks (e.g. `std::hermite` at first, then `std::laguerre`, etc.) might make sense as well (also see note on boost.math below).

I started out from this abandoned merge request: https://reviews.llvm.org/D58876 .

The C++23 standard defines them in-terms of `/* floating-point type */` arguments. I have not looked into that.

Note, there is still an ongoing discussion on discourse whether importing boost.math is an option.

PaulXiCao commented:
@mordante I would be thankful for feedback on this implementation.

Zingam commented:
You should prefix your PR with something like [libc++] or [libc++][math], etc.

Zingam commented on libcxx/include/experimental/math:
```diff
@@ -0,0 +1,45 @@
+#if _LIBCPP_STD_VER >= 17
+
+#  include <experimental/__math/hermite.h>
+#  include <type_traits> // enable_if_t, is_integral_v
```
I don't do reviews usually due to my incompetence but here are my 2 cents.

I believe we do includes unconditionally. We also include the granularized headers e.g. enable_if.h, instead of type_traits? Or is it required by the standard to include it?

Zingam replied:
Thanks!

@PaulXiCao I'd recommend to check recent PRs and their comments as a general guide for style and recent best practices.

PaulXiCao replied:
Good idea! Will do so.

mordante commented on libcxx/include/experimental/__math/hermite.h:
```diff
@@ -0,0 +1,49 @@
+/// Press, William H., et al. Numerical recipes 3rd edition: The art of
+/// scientific computing. Cambridge university press, 2007, p. 183.
+template <class _Real>
+_Real __libcpp_hermite_recurrence(const unsigned __n, const _Real __x) {
```
Why did you make the arguments const qualified?

PaulXiCao replied:
I usually follow the philosophy that one should try to use `const` wherever possible. To simplify reasoning about code.
You mentioned somewhere that `const` is used seldomly in libcxx. Thus, I removed them.

mordante reviewed (commented):
Thanks a lot for working on this!
I mainly focused on the libc++ specific parts of the implementation.

Zingam commented on libcxx/include/experimental/math:
```diff
@@ -0,0 +1,45 @@
+
+inline _LIBCPP_HIDE_FROM_ABI long double hermitel(unsigned __n, long double __x) { return hermite(__n, __x); }
+
+template < class _Integer >
```
```suggestion
template <class _Integer>
```

Please note that templates are untouched by the formatter but this is how you should format "modern" templates.

mordante replied:
Correct due to C++03 compatibility we do not change spaces. Closing elements like `>>` are not allowed in C++03. There `> >` is required.

Zingam replied:
Yeah, I wanted to say something like "without spaces" but failed to do it clearly. Thanks for correcting.

PaulXiCao replied:
Fixed in d4debd5f7751154cca1271c7f6a0c0979ed0c52b.

PaulXiCao commented on libcxx/include/CMakeLists.txt:
```diff
@@ -922,6 +922,7 @@ set(files
   execution
   expected
   experimental/__config
+  experimental/__math/hermite.h
```
Ok, I moved the code to the `include/cmath` header. 
Is there a policy about adding new folders, files? If so how granular can we go? E.g. could I create the `include/__cmath/` folder and `include/__cmath/hermite.h` file? Or should it rather be one header for all special math functions, i.e. `include/__cmath/special_functions.h`?
On the same note, one could factor out `std::lerp` functions from `include/cmath` into `include/__cmath/lerp.h`.

PaulXiCao replied:
See [comment below](https://github.com/llvm/llvm-project/pull/89982#discussion_r1581909160).

PaulXiCao commented on libcxx/include/experimental/__math/hermite.h:
```diff
@@ -0,0 +1,49 @@
+/// \file
+/// This file contains the internal implementations of std::hermite.
+///
+//===----------------------------------------------------------------------===//
```
Fixed in a85b2bc71c075137926978ad5ab464dcd77f9171.

PaulXiCao replied:
I am not sure if it is the preferred workflow here but I will close threads on my own (after implementing a fix)?!

PaulXiCao commented on libcxx/include/experimental/__math/hermite.h:
```diff
@@ -0,0 +1,49 @@
+/// Press, William H., et al. Numerical recipes 3rd edition: The art of
+/// scientific computing. Cambridge university press, 2007, p. 183.
+template <class _Real>
+_Real __libcpp_hermite_recurrence(const unsigned __n, const _Real __x) {
```
Fixed in be6b88c.

PaulXiCao commented on libcxx/include/experimental/__math/hermite.h:
```diff
@@ -0,0 +1,49 @@
+
+#include <cmath> // std::isnan
+#include <experimental/__config>
+#include <limits> // std::numeric_limits
```
Fixed.

PaulXiCao commented on libcxx/test/std/experimental/math/hermite.pass.cpp:
```diff
@@ -0,0 +1,303 @@
+
+#include <assert_macros.h>
+
+// UNSUPPORTED: c++03, c++11, c++14
```
Fixed in e2e2da87bf5b5ade98eba92946eb23804fc7fa5a.

PaulXiCao commented on libcxx/include/experimental/__math/hermite.h:
```diff
@@ -0,0 +1,49 @@
+template <class _Real>
+_Real __libcpp_hermite_recurrence(const unsigned __n, const _Real __x) {
+  if (0 == __n)
+    return 1;
```
Fixed in c9fa9b4fdd63d1feb94e1d5fc29145649df01107.

PaulXiCao commented on libcxx/include/experimental/__math/hermite.h:
```diff
@@ -0,0 +1,49 @@
+  _Real __H_nPrev{1};
+  _Real __H_n = 2 * __x;
+  for (unsigned __i = 1; __i < __n; ++__i) {
+    const _Real __H_nNext = 2 * (__x * __H_n - __i * __H_nPrev);
```
Fixed in 947071a.

PaulXiCao commented on libcxx/include/experimental/__math/hermite.h:
```diff
@@ -0,0 +1,49 @@
+/// Press, William H., et al. Numerical recipes 3rd edition: The art of
+/// scientific computing. Cambridge university press, 2007, p. 183.
+template <class _Real>
+_Real __libcpp_hermite_recurrence(const unsigned __n, const _Real __x) {
```
Fixed in 40404493dc222d161dd46853a295d2117b1d137b. (Just copied from previously abandoned pr.)

Zingam commented:
@PaulXiCao It may be useful to add a status page to track the progress and assignments.

PaulXiCao commented on libcxx/include/experimental/math:
```diff
@@ -0,0 +1,45 @@
+
+_LIBCPP_BEGIN_NAMESPACE_EXPERIMENTAL
+
+inline _LIBCPP_HIDE_FROM_ABI double hermite(unsigned __n, double __x) { return __libcpp_hermite(__n, __x); }
```
Fixed in 925349eced1575b1d3fbe1442b68ffc79d4f9f9b.

philnik777 commented on libcxx/utils/libcxx/header_information.py:
```diff
@@ -56,6 +56,7 @@
     "cwchar": "// UNSUPPORTED: no-wide-characters",
     "cwctype": "// UNSUPPORTED: no-wide-characters",
     "experimental/iterator": "// UNSUPPORTED: c++03",
+    "experimental/math": "// UNSUPPORTED: c++03, c++11, c++14",
```
Is this a leftover?

PaulXiCao replied:
Yes. Fixed in fc667d6e28dc4b4ec139d42cc36bb2b1414f879a.

philnik777 commented on libcxx/docs/Status/Cxx17Papers.csv:
```diff
@@ -26,7 +26,7 @@
 "","","","","",""
 "`P0024R2 <https://wg21.link/P0024R2>`__","LWG","The Parallelism TS Should be Standardized","Jacksonville","|Partial|",""
-"`P0226R1 <https://wg21.link/P0226R1>`__","LWG","Mathematical Special Functions for C++17","Jacksonville","",""
+"`P0226R1 <https://wg21.link/P0226R1>`__","LWG","Mathematical Special Functions for C++17","Jacksonville","|In Progress|",""
```
Please add a note on what is already implemented.

PaulXiCao replied:
I added a status page as asked for in [comment above](https://github.com/llvm/llvm-project/pull/89982#issuecomment-2080161608). Implemented in 18a320a0f523d1939262d1cfc633da6e5fae4870.

PaulXiCao replied:
I suppose that we do not need to duplicate that info? Should we link the status page therein though?

philnik777 commented on libcxx/include/cmath:
```diff
@@ -765,6 +766,54 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp __constexpr_scalbn(_Tp _
+
+template <class _Integer>
+_LIBCPP_HIDE_FROM_ABI std::enable_if_t<std::is_integral_v<_Integer>, double> hermite(unsigned int __n, _Integer __x) {
+  // use double internally -- see C++17 standard - 29.9.1.2.2
```
No need to call this out specifically here. It's the same for all the math functions.

PaulXiCao replied:
Fixed in ca87828caa9ce8ed32203f9ddd9e9203c3e630d9.

philnik777 commented on libcxx/include/cmath:
```diff
@@ -765,6 +766,54 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp __constexpr_scalbn(_Tp _
+
+inline _LIBCPP_HIDE_FROM_ABI long double hermitel(unsigned __n, long double __x) { return std::hermite(__n, __x); }
+
+template <class _Integer>
+_LIBCPP_HIDE_FROM_ABI std::enable_if_t<std::is_integral_v<_Integer>, double> hermite(unsigned int __n, _Integer __x) {
```
```suggestion
template <class _Integer, std::enable_if_t<std::is_integral_v<_Integer>, int> = 0>
_LIBCPP_HIDE_FROM_ABI double hermite(unsigned int __n, _Integer __x) {
```

PaulXiCao replied:
I do not have too much experience with TMP. Could you elaborate on the advantage of your proposed change? Introducing an unused template argument type `int` seems strange to me.

PaulXiCao replied:
Just found other code which also uses the style you suggested (e.g. [libcxx/include/__math/abs.h#L37](https://github.com/llvm/llvm-project/blob/352602010fac5c7e07bc6b992848ab746d0c2857/libcxx/include/__math/abs.h#L37)). I will implement the change you suggested, but would still welcome your thoughts on this :)

PaulXiCao replied:
Fixed in 51f82a50b9007edbb000cfcdcc353dfc28cb8ab2.

philnik777 replied:
There's nothing inherently wrong about return type SFINAE. It's just that the `enable_if_t<..., int> = 0` version works everywhere. Return type SFINAE doesn't work on constructors, `enable_if_t<...>* = nullptr` doesn't work in C++03 etc. This avoids confusion on which version should be used in which case.

PaulXiCao replied:
Is `std::is_integral` the correct concept to check? I am wondering as it would also allow types that seem unreasonable/indicative of wrong usage, e.g. `bool`.
From [cppreference.com](https://en.cppreference.com/w/cpp/numeric/special_functions/hermite) ("Notes" section): _"They only need to be sufficient to ensure that for their argument num of integer type, std::hermite(int_num, num) has the same effect as std::hermite(int_num, static_cast<double>(num))."_
From the C++23 Standard
* 28.7.1.3: _"...where arguments of integer type are considered to have the same floating-point conversion rank as double. ..."_. This defines that `std::hermite(n,x)` should also accept integer values for its `x` argument. 
* Integer types are defined in 6.8.2. and do not contain `bool`.

philnik777 commented on libcxx/include/cmath:
```diff
@@ -765,6 +766,54 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp __constexpr_scalbn(_Tp _
+  return __H_n;
+}
+
+inline _LIBCPP_HIDE_FROM_ABI double hermite(unsigned __n, double __x) { return std::__hermite(__n, __x); }
```
Not attached: should the mathematical special functions be in the dylib? I expect this will be quite a bit of code, and it would be nice to be consistent between the special functions.

PaulXiCao replied:
I cannot anticipate how much code that will be nor what the threshold is when we should be using dynamic libraries. Could we make that switch with a later integration, e.g. once the LOC exceed some threshold?

philnik777 replied:
Moving code into the dylib has quite a few consequences and may change design considerations, so I don't think we should wait until some later point. Unless we want to make the functions experimental for now.

PaulXiCao replied:
Ok, I will need to look into that. Can you point me in some direction? Documentation or other sample code?

PaulXiCao replied:
From a discussion with @mordante: 

* This function is quite small a probably does not need to go into the dylib.
* Large code or special template magic could be moved there.
* Concerning the other special math functions: I assume they probably will not contain too much code nor  any special template magic. 
* We still have the option in the future to move other special math functions into the dylib even though this one might stay in the header.

philnik777 commented on libcxx/include/cmath:
```diff
@@ -765,6 +766,54 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp __constexpr_scalbn(_Tp _
+template <class _Real>
+_LIBCPP_HIDE_FROM_ABI _Real __hermite(unsigned __n, _Real __x) {
+  if (std::isnan(__x))
+    return std::numeric_limits<_Real>::quiet_NaN();
```
Is this correct? I'm not sure how to read the standard here, but it'd make a lot more sense to me if it returned the same NaN as was passed into the function.

PaulXiCao replied:
For reference, the [Standard 28.7.6.1.1](https://wg21.link/sf.cmath.general): "_If any argument value to any of the functions specified in [sf.cmath] is a NaN (Not a Number), the function shall return a NaN but it shall not report a domain error. ..._"

I would interpret this as we must return a NaN but the actual NaN type is implemenation-defined?!

But as you point it out, I thought some more about it and I might also prefer the straight-forward alternative: `if (isnan(x)) return x;`.

Note, that the initial version (older abandoned mr) also contained the conversion.

I had a quick peed into my local libstdc++ installation. It also uses the conversion to `quiet_NaN`. Which makes we wonder if there is a deeper reason to it?

lntue replied:
+1 on `return x`, so that signaling NaNs are propagated on supported platforms.

PaulXiCao replied:
Fixed in 7f003acbf6ea08551450c4b1bb3024b2b8635ded.

philnik777 commented on libcxx/include/cmath:
```diff
@@ -765,6 +766,54 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp __constexpr_scalbn(_Tp _
   return __builtin_scalbn(__x, __exp);
 }
 
+#if _LIBCPP_STD_VER >= 17
```
This should probably be in `include/__math/special_functions.h`.

PaulXiCao replied:
I raised similar thoughts in a [previous comment](https://github.com/llvm/llvm-project/pull/89982#discussion_r1581533949). 
Is it valid to use the `__math/` folder instead of a newly created `__cmath/` folder? Note, that the special math functions are only part of the `<cmath>` and not `<math.h>` header. 
In general, I am more than in favor to move that code into a separate header.

philnik777 replied:
Yes, that's fine. It's been named `__math` instead of `__math_h` so it makes sense to move any math code in there.

philnik777 commented on libcxx/include/cmath:
```diff
@@ -765,6 +766,54 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp __constexpr_scalbn(_Tp _
+  if (std::isnan(__x))
+    return std::numeric_limits<_Real>::quiet_NaN();
+
+  _Real __H_0{1};
```
Why does the standard make `n >= 128` implementation-defined?

PaulXiCao replied:
I could only speculate. Not sure where to find information on that...

philnik777 replied:
Probably somewhere deep within the archives of wg21. I hope @lntue has some insight here.

lntue replied:
I have no idea why they put that in the standard.  My guess is that some might implement the recurrence relation without multiply-by-2 part, and then scale the final results by `2^n`.  For `n >= 128`, `2^128` exceeds the range of `float` type, so this way of computing hermite polynomials might not support `n >= 128` for `float`.

MattStephanson replied:
See https://wg21.link/n1884. TLDR: the function changes so rapidly that it's difficult to even get the correct sign.

philnik777 commented on libcxx/include/cmath:
```diff
@@ -765,6 +766,54 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp __constexpr_scalbn(_Tp _
+    return std::numeric_limits<_Real>::quiet_NaN();
+
+  _Real __H_0{1};
+  if (__n == 0)
```
Not attached: You need to add a section in `libcxx/docs/ImplementationDefinedBehavior.rst` for the aforementioned implenentation-defined behaviour.

PaulXiCao replied:
Fixed in 3b947d166c12ec210bbb448c4fd69fac9e7efc10.

PaulXiCao commented:
> @PaulXiCao It may be useful to add a status page to track the progress and assignments. https://libcxx.llvm.org/#getting-started-with-libc

Done in 18a320a0f523d1939262d1cfc633da6e5fae4870.

philnik777 commented on libcxx/docs/ImplementationDefinedBehavior.rst:
```diff
@@ -51,6 +51,17 @@ Libc++ determines that a stream is Unicode-capable terminal by:
+
+Most functions within the Mathematical Special Functions section contain integral indices.
+The Standard specifies the result for larger indices as implementation-defined.
+Libc++ pursuits reasonable results by choosing the same formulas as for indices below that threshold.
```
Have you checked whether the results actually make sense beyond 128?

PaulXiCao replied:
I only check upto `n<128`. 
By the nature of it we can only check upto some finite value and this given value from the Standard seemed reasonable. Also note that our implementation does not distinguish between different orders (e.g. nothing like below/above any threshold). Do you think that we should increase the test coverage for the parameters?

philnik777 commented on libcxx/docs/Status/SpecialMathPapers.csv:
```diff
@@ -0,0 +1,2 @@
+Number,Name,Standard,Assignee,Status,First released version
+`P0226 <https://wg21.link/P0226>`_,"Mathematical Special Functions for C++17","C++17",,"|In Progress|"
```
You don't need a list of papers if the status page is for a specific one.

PaulXiCao replied:
Fixed in 7d8c06d72513ed293d731915dfbb50dd001f912d.

philnik777 commented on libcxx/docs/Status/SpecialMathProjects.csv:
```diff
@@ -0,0 +1,22 @@
+| `[sf.cmath.ellint.2] <https://wg21.link/sf.cmath.ellint.2>`_, std::ellint_2, None, |Not Started|
+| `[sf.cmath.ellint.3] <https://wg21.link/sf.cmath.ellint.3>`_, std::ellint_3, None, |Not Started|
+| `[sf.cmath.expint] <https://wg21.link/sf.cmath.expint>`_, std::expint, None, |Not Started|
+| `[sf.cmath.hermite] <https://wg21.link/sf.cmath.hermite>`_, std::hermite, None, |Complete|
```
Maybe add yourself as the assignee?

PaulXiCao replied:
Done in 62841f461e4c913fcdf43fa737757b1b36d8e3c2.

philnik777 commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,310 @@
+            T(5.3874808900112)};
+
+  default:
+    throw "Roots of Hermite polynomial of order " + std::to_string(n) + " not implemented!\n";
```
This will break with exceptions disabled. Just `assert(false)`. Actually, why is this not just an array? This seems like a very weird way to have a list of data.

PaulXiCao replied:
Replaced by `assert(0)` in 3ddac0b30dd81aae5482b759de59be269678d36c, and changed to `assert(false)` in 293dd2ba1c8a9f88aadddfb36936ba7d6f01ddc2.
A clang on Windows [job](https://github.com/llvm/llvm-project/actions/runs/9169528013/job/25210556931?pr=89982) issues an compile error: "error: non-void function does not return a value in all control paths [-Werror,-Wreturn-type]". Appended a `return {};` in cc58886185ec8e9488133ef9e40b838e14f1f2a5.

Still looking for a better solution though.

> Actually, why is this not just an array?

The number of return values depends on the input and I also want to loop over them, e.g.
```cpp
for (int n=0;n<20;++n) 
  for (T x : get_roots<T>(n)) 
    ...
```
I do not find it straight-forward to implement via arrays. Please reach out if you have already something in mind how to improve this here.

philnik777 commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,310 @@
+
+#include <assert_macros.h>
+
+namespace {
```
Generally, we don't have anonymous namespaces in the tests. It's not like we get much of a benefit from it.

PaulXiCao replied:
Fixed in 71dcd12a17d321d46af15f13e7f9d3ee02604940.

philnik777 commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,310 @@
+} // namespace
+
+int main(int, char**) {
+  test_hermite<float>(1e-5f, 1e-5f);
```
Please refactor the tests to the style
```c++
template <class T>
void test() {
  { // Write what the test checks
    // put the test here
  }

  { // and the next test
    // some code
  }
}
```
We generally use this in new tests.

PaulXiCao replied:
Do I understand you correctly that the `main` should only call 3 `test<T>();` functions (for `T=float, double, long double`)?

PaulXiCao replied:
Implemented in 8b57a634c764c2aef653106fbb6b762663f9cf75.

philnik777 commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,310 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
```
Please add the synopsis of what this is testing.

PaulXiCao replied:
Fixed in 976309af4d8b289d2b55d9b68437203ed5536618.

philnik777 commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,310 @@
+#include <vector>
+#include <array>
+
+#include <assert_macros.h>
```
This looks unused?

PaulXiCao replied:
Fixed in b42cd04c8b872c0d05fbbf81824c08ffc7103220.

philnik777 commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,310 @@
+
+// UNSUPPORTED: c++03, c++11, c++14
+
+#include <cassert>
+#include <cmath>
+#include <limits>
+#include <vector>
+#include <array>
```
```suggestion
#include <array>
#include <cassert>
#include <cmath>
#include <limits>
#include <vector>
```

PaulXiCao replied:
Fixed in 840321277748f8d62682e6d92420e0b1ac39848e.

lntue commented on libcxx/include/cmath:
```diff
@@ -765,6 +766,53 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp __constexpr_scalbn(_Tp _
+    __H_n_prev       = __H_n;
+    __H_n            = __H_n_next;
+  }
+  return __H_n;
```
I think we could have somewhat accurate overflow detection here with:
```
if (std::isnan(__H_n)) {
  // Overflow.
  if (n & 1)
    // n is odd, return infinity of the same sign as x.
    return std::copysign(std::numeric_limits<_Real>::infinity(), x);
  // n is even, return +Inf
  return std::numeric_limits<_real>::infinity();
}
return __H_n;
```

PaulXiCao replied:
While I was thinking about your comment for quite some time multiple issues crossed my mind.

* I assume your point is that in the nature of subtraction and recursive calls (e.g. `_Real __H_n_next = 2 * (__x * __H_n - __i * __H_n_prev);`) one could produce an Inf of the wrong sign?
  * I will try to find a test set which would show this issue for the current implementation.
* I believe one needs to clearly separate between NaNs and Infs.
  * NaNs are already handled at function entry. They should not be producible within this function.
  * The check in your code snippet should probably test for Infs, i.e. `if (std::isinf(__H_n)) { ... }`?
* This does clutter the implementation quite a bit (mixing error checks with formula). I assume similar logic will need to apply for some of the other special math functions. We might probably want to extract that logic at some point.

lntue replied:
+1 to everything in your comment.
- For the first point a simple non-trivial (might be not that useful) example is when `x = std::numeric_limits<_Real>::max()`, following the iterations, you will get that:
```
  hermite(0, x) = 1.0
  hermite(1, x) = +Inf
  hermite(2, x) = +Inf
  hermite(3, x) = NaN
  ...
```
and there are definitely more intricate cases.
- Another issue that we might want to address later is precision loss.  I haven't looked into the stability of the iterations, but we might have catastrophic cancellation when `__x * __H_n ~ __i * __H_n_prev`.  It is interesting to find some non-trivial examples, but we can defer dealing with this case to later PR.

PaulXiCao replied:
We could make use of your proposed code snippet at the entry of the function, e.g.
```cpp
_LIBCPP_HIDE_FROM_ABI _Real __hermite(unsigned __n, _Real __x) {
  if (std::isnan(__x))
    return __x;
  if (std::isinf(__x)) {
    // ... your proposed overflow handling
  }
  // ...
```
Though I am unsure of the importance of handling `std::hermite(n, +-Inf)` correctly. Personally I would not expect correct output for such an input (informally known as "garbage in, garbage out" ;D).
Let me know if this is worth implementing.

lntue replied:
I'm thinking of something like this:
https://godbolt.org/z/7MaYYzob9

PaulXiCao replied:
Looks good. Can I ask why you check for `std::isnan` instead of `std::isinf`? I believe I have seen examples where at some `n` the next 2 results were `Inf` and then turned to `NaN`. If I find some more time I will try to find such an example.

lntue replied:
So I think for finite `__x`, if some intermediate `__H_n` is `Inf`, it is "correctly" overflowed (I might need to prove it or if it is not right, find some counter example).  So checking `isnan` should be enough.  OTOH, probably we should just use `!std::isfinite(__H_n)` instead.

PaulXiCao replied:
I looked some more into possible overflows. First thing to notice is that double overflows at ~1e308.
Hermite polynomial H_127 does not reach such large values inbetween any roots but only for larger x:
* largest root is at x=~16
* function values previous to the last root are significantly below the overflow threshold, i.e. |H_127(x)| < ~1e190 for |x| < ~16
* overflow threshold is reached at roughly: H_127(x=140)~1e308

For polynomials of smaller order it is even more extreme (e.g. largest root is smaller and overflow happens at even larger x).
I would not take polynomials of higher order into this discussion as their result is deemed implementation-defined by the Standard. Thus, possibly slowing down code for n<128 because of arguments for n>=128 seems strange.

My conclusion: I believe your initial suggestion is enough (e.g. checking the value of `__H_n` *after* the loop). Using `!std::isfinite()` also seems the approriate check here.

I will probably push a change similar to `hermite2` as in this [godbolt](https://godbolt.org/z/6xojME18W). (`hermite` is the current implementation, and `hermite1` the one you proposed in the previous godbolt.)

PaulXiCao replied:
Implemented in 511b45679ab65dc3efca81abc976c27711a5df9a.

lntue replied:
SGTM

lntue commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,310 @@
+
+template <class T>
+std::array<T, 7> sample_points() {
+  return {-1.0, -0.5, -0.1, 0.0, 0.1, 0.5, 1.0};
```
Can you add some sample points that have magnitude `> 1` and `+- Inf`?

PaulXiCao replied:
Yes, makes sense. I'll look into that.

PaulXiCao replied:
Fixed in 0098ee77a763ea393f0993037e9d93aab2a209c9 (for points with magnitude `>1`) and c94fad9ba48a4bb4149ab5cc8f97e1a2a3ac34f4 (for `+-Inf`).

mordante commented on libcxx/docs/Status/SpecialMath.rst:
```diff
@@ -0,0 +1,35 @@
+.. special-math-status:
```
As mentioned in our call this file needs to be added to https://github.com/llvm/llvm-project/blob/main/libcxx/docs/index.rst?plain=1

PaulXiCao replied:
Fixed in 84b7b87e968831265f000f008016fbfd3473f258.

mordante commented on libcxx/include/cmath:
```diff
@@ -315,6 +315,7 @@ constexpr long double lerp(long double a, long double b, long double t) noexcept
 #include <limits>
 #include <version>
 
+#include <__math/special_functions.h>
```
As mentioned in our call the synopsis in this file needs to list the added hermite functions.
These functions also need to be uncommented in https://github.com/llvm/llvm-project/blob/main/libcxx/modules/std/cmath.inc

PaulXiCao replied:
6b28fc59cc674f15075ec058ac6e3b1b217747bc: uncommented in cmath.inc
58ff841509fefa04784d788af4b8e6f18c84d5dd: added to synopsis in cmath

H-G-Hristov commented on libcxx/include/__math/special_functions.h:
```diff
@@ -0,0 +1,73 @@
+
+#if _LIBCPP_STD_VER >= 17
+
+/// \return the Hermite polynomial \f$ H_n(x) \f$
+/// \note The implementation is based on the recurrence formula
+/// \f[
+///   H_{n+1}(x) = 2x H_n(x) - 2n H_{n-1}
+/// \f]
+/// Press, William H., et al. Numerical recipes 3rd edition: The art of
+/// scientific computing. Cambridge university press, 2007, p. 183.
+template <class _Real>
+_LIBCPP_HIDE_FROM_ABI _Real __hermite(unsigned __n, _Real __x) {
```
I believe there was a comment already that we don't use doxygen comments. I think you should move that comment inside the function body.

PaulXiCao replied:
Fixed in d4ea6088dca10dde0cfc21ad25ec3df218955c8c.

mordante commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,312 @@
+
+#include "type_algorithms.h"
+
+inline constexpr unsigned MAX_N = 128;
```
This is not a function or a static data member, so it's not implicit `inline`.

For templates we use both with and without `inline` and there is not strong preference.

For functions some people have a strong preference to use `inline constexpr`; and the `inline` is redundant.

Typically we don't use capitalized constants.

PaulXiCao replied:
Renamed the constant to `g_max_n` in df4eadfa73f14831d6cd1bd045d5f5081dea31d6. Unsure if this is the correct naming scheme as I did not find another file with an example. Let me know if it should be changed to something else.

mordante commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -238,78 +196,117 @@ std::vector<T> get_roots(unsigned n) {
 
   default:
-    throw "Roots of Hermite polynomial of order " + std::to_string(n) + " not implemented!\n";
+    assert(0);
```
```suggestion
    assert(false);
```

PaulXiCao replied:
Fixed in 293dd2ba1c8a9f88aadddfb36936ba7d6f01ddc2.

PaulXiCao commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,312 @@
+//    float       hermitef(unsigned n, float       x);
+//    long double hermitel(unsigned n, long double x);
+//    template <class Integer>
+//    double      hermite( unsigned n, Integer     x);
```
Fixed in 67b0da98f2a62422683135a61cd4e14f56770582.

PaulXiCao commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,312 @@
+  case 0:
+    return {};
+  case 1:
+    return {T(0)};
```
I want to use the same style on this line as in the following ones. And the next lines need to perform a narrowing conversion (e.g. `return {T(0.707106781186548)};` for `T=float`). Using braces I assumed that the compiler or some tool (e.g. clang-tidy) would throw warnings. Let me know if this is baseless.

mordante reviewed (commented):
@lntue Thanks for your help reviewing! Can you approve the patch when you feel it's correct from the math point of view? I'll look at the libc++ specifics.

The patch seems mostly good to me.

lntue reviewed (approved):
The math part looks good to me!  Thanks!

PaulXiCao commented on libcxx/test/std/numerics/c.math/hermite.pass.cpp:
```diff
@@ -0,0 +1,333 @@
+  }
+
+  { // check input infinity is handled correctly
+    Real inf = std::numeric_limits<Real>::infinity();
```
Implemented in 55ec209c8a261e8a48cbefc7a63e0ca776905b90. This contains only tests for `T=double`. It is dependent on the size of the type. `long double` can be different on other platforms which would result in quite messy test code. Let me know if I should generalize this test some more.

efriedma-quic commented:
Should the documentation record the known error in ulps of the computed result?  See, for comparison, https://libc.llvm.org/math/index.html .

PaulXiCao commented:
> Should the documentation record the known error in ulps of the computed result? See, for comparison, https://libc.llvm.org/math/index.html .

The cited page lists functions (e.g. `sin`) vs input type (e.g. `float`) and their correctness, i.e. errors counted in ULP for each rounding mode.

For any non-trivial implementation this is mostly calculated by comparing the result with a reference implementation, e.g. an implementation using larger floating-point types or a multi-precision library.

This worst-case analysis only holds any meaning for well chosen sample points. We would need to decide what a common input range is.

All of this is quite tedious and complicated. Which is why it probably has not been done for most of the "Higher Math Functions" in libc? (The table is almost empty.)

A plain documentation by creating a similar table for the "Mathematical Special Functions" only filled with "N/A" does not yield any benefit from my point of view.

efriedma-quic commented:
The empty squares in that table are stuff that hasn't been implemented in llvm-libc yet.  (llvm-libc is, in general, currently a work in progress.)  glibc also has a similar table in their documentation.

For "float" inputs, in some cases you can probably figure out the worst-case error across the whole range by brute force.  For wider inputs, you'd need some combination of random sampling, and brute-force of a few carefully chosen small intervals, yes.

PaulXiCao commented:
There is also another option: We could consider splitting off the correctness documentation into its own issue/merge request.

My thoughts:
- There is no real reason why everything must be handled in one pr.
- Correctness documentation seems to be a more involved issue which will probably need quite some discussion.
- There are other pr waiting for this integration (e.g. `std::legendre`).
- On the flip-side: There is the possibility that it might be pushed to some backlog and not get fixed in the foreseeable future.

mordante commented:
> There is also another option: We could consider splitting off the correctness documentation into its own issue/merge request.
> 
> My thoughts:
> 
>     * There is no real reason why everything must be handled in one pr.
> 
>     * Correctness documentation seems to be a more involved issue which will probably need quite some discussion.
> 
>     * There are other pr waiting for this integration (e.g. `std::legendre`).
> 
>     * On the flip-side: There is the possibility that it might be pushed to some backlog and not get fixed in the foreseeable future.

In libc++ we have not added this documentation for other math functions. For example `lerp`. so I don't feel this should be a blocker for this patch.

It would be nice to do in a separate patch.

PS I just restarted to build.

efriedma-quic commented:
I'm less concerned about the documentation, and more concerned that you've done the underlying analysis to prove the implementation actually has the numerical properties you want over the defined range.  If you skip that analysis, you end up with issues like #92782.

mordante commented:
> I'm less concerned about the documentation, and more concerned that you've done the underlying analysis to prove the implementation actually has the numerical properties you want over the defined range. If you skip that analysis, you end up with issues like #92782.

Fair point. I'm not too well versed in the math part. @lntue looked at the math part of the implementation.

mordante reviewed (approved):
LGTM, thanks a lot for your work @PaulXiCao. Can you rebase the patch on main to give the CI one more run. When that passes I'll merge the patch.

PaulXiCao commented:
That sounds great! I will rebase in the coming days.

PaulXiCao commented:
@mordante I have just rebased to the current main branch.

mordante commented:
The CI failure is unrelated to your patch. I wanted to commit this patch but I noticed your e-mail address is hidden. Could you make this visible?

PaulXiCao commented:
@mordante My email address is hidden on purpose. Is it really required?

PaulXiCao committed:
apply original diff from D58876

This commit was created by taking the abandoned merge request from
https://reviews.llvm.org/D58876 .

PaulXiCao committed:
cmake: fix include by adding new files to file listing

PaulXiCao committed:
moved new files into math/

PaulXiCao committed:
replace "_VSTD" by "std". similar to D117811

see https://reviews.llvm.org/D117811

PaulXiCao committed:
replace deprecated _LIBCPP_INLINE_VISIBILITY by _LIBCPP_HIDE_FROM_ABI

PaulXiCao committed:
move test files from internal libcxx/ into std/

PaulXiCao committed:
legendre tests succeeds! (fixed compiler warning: shadowing variable)

PaulXiCao committed:
laguerre tests succeeds! (fixed compiler warning: shadowing variable)

PaulXiCao committed:
hermite tests succeed! (fixed compiler warning: shadowing variable)

PaulXiCao committed:
assoc_laguerre tests succeeds! (fixed compiler warning: shadowing variable)

PaulXiCao committed:
hermite: fix documentation comment

PaulXiCao committed:
assoc_legendre tests succeeds! (fixed compiler warning: shadowing variable)

PaulXiCao committed:
set header info for lit

PaulXiCao committed:
create experimental/math module

PaulXiCao committed:
hermite: minor code changes: rename variables, brace initialization

PaulXiCao committed:
use more readable variable names

PaulXiCao committed:
remove explicit type conversions

PaulXiCao committed:
keeping only hermite, i.e. remove laguerre, legendre

PaulXiCao committed:
make use of function template argument deduction

PaulXiCao committed:
unify return path via ternary op

PaulXiCao committed:
constness of passed arguments

PaulXiCao committed:
clang format

PaulXiCao committed:
sufficient additional overloads by LWG 3234

PaulXiCao committed:
cleanup preprocessor usage

PaulXiCao committed:
remove old comment

PaulXiCao committed:
simplify hermite(float)

PaulXiCao committed:
implement: template<class Integer> double hermite(unsigned, Integer)

PaulXiCao committed:
cleanup tests

PaulXiCao committed:
update docs/Status

PaulXiCao committed:
make use of Lit: // UNSUPPORTED: c++XX

PaulXiCao committed:
move out of experimental/

PaulXiCao committed:
modulemap include private headers

PaulXiCao committed:
remove doxygen style file comment

PaulXiCao committed:
include missing header: type_traits/is_integral

PaulXiCao committed:
all functions need _LIBCPP_HIDE_FROM_API

PaulXiCao committed:
remove `std::` similar to other surrounding code

PaulXiCao committed:
test: place Lit comment correctly

PaulXiCao committed:
don't use yoda conditional

PaulXiCao committed:
edits according to mentioned style issues

PaulXiCao committed:
style format of template argument

PaulXiCao committed:
shorten internal hermite function's name

PaulXiCao committed:
fully qualify function calls

PaulXiCao committed:
status page

PaulXiCao committed:
remove dead code: lit header comments

PaulXiCao committed:
test: style changes, renaming, etc.

PaulXiCao committed:
remove comment about integer->double conversion

PaulXiCao committed:
specify implementation-defined behavior: hermite(n,x) for n>=128

PaulXiCao committed:
hermite(unsigned, _Integer): changed SFINAE usage

PaulXiCao committed:
link status page from general C++17 status page

PaulXiCao committed:
move implementation from include/cmath to include/__math/special_functions.h

PaulXiCao committed:
simplify NaN handling (no conversion)

PaulXiCao committed:
remove anonymous namespace in test file

PaulXiCao committed:
simplify status documentation: single paper needs no table

PaulXiCao committed:
assign std::hermite to myself

PaulXiCao committed:
test: remove unused include

PaulXiCao committed:
test: add synopsis

PaulXiCao committed:
test: sort includes

PaulXiCao committed:
test: refactor/style change. only test<T>() within main()

PaulXiCao committed:
test: replace exception throwing by assert

PaulXiCao committed:
test: make use of types::for_each

PaulXiCao committed:
test: checks for x=+-inf as input

PaulXiCao committed:
handle overflows

PaulXiCao committed:
test: additional sample points with abs(x)>1

PaulXiCao committed:
cmath.inc: uncomment std::hermite functions

PaulXiCao committed:
cmath: added implemented functions to synopsis

PaulXiCao committed:
index.rst: add special math status page to TOC

PaulXiCao committed:
cmath: add hermite{l,f} to synopsis

PaulXiCao committed:
get rid of doxygen-style comments

PaulXiCao committed:
test: format test's synopsis

PaulXiCao committed:
test: inf result for finite input

PaulXiCao committed:
test: assert(false) instead of assert(0)

PaulXiCao committed:
test: renamed MAX_N to g_max_n

PaulXiCao committed:
test: add more values for integeral x

PaulXiCao committed:
rename to use ugly name: inf -> __inf

PaulXiCao committed:
silence clang-tidy warning about variable name in math code

PaulXiCao committed:
Revert "export functions for module usage"

This reverts commit 165fc118554d4be52898020fbe6edf108e41a217.

PaulXiCao committed:
Adds a work-around to pass the module tests.

PaulXiCao committed:
export functions for module usage

PaulXiCao committed:
fix test: unsupported case must also return a value

PaulXiCao commented:
> The CI failure is unrelated to your patch. I wanted to commit this patch but I noticed your e-mail address is hidden. Could you make this visible?

@mordante I added my email to each commit.

mordante commented:
> > The CI failure is unrelated to your patch. I wanted to commit this patch but I noticed your e-mail address is hidden. Could you make this visible?
> 
> @mordante I added my email to each commit.

Thanks! I see one Apple build failure. I'm not sure it's related so I started to other CI runners.

mordante commented:
> > > The CI failure is unrelated to your patch. I wanted to commit this patch but I noticed your e-mail address is hidden. Could you make this visible?
> > 
> > 
> > @mordante I added my email to each commit.
> 
> Thanks! I see one Apple build failure. I'm not sure it's related so I started to other CI runners.

The other builds pass. Could you rebase to make sure the apple build is indeed a fluke. The number of errors it gives makes me uncomfortable to merge without a new run.

PaulXiCao commented:
> > > > The CI failure is unrelated to your patch. I wanted to commit this patch but I noticed your e-mail address is hidden. Could you make this visible?
> > > 
> > > 
> > > @mordante I added my email to each commit.
> > 
> > 
> > Thanks! I see one Apple build failure. I'm not sure it's related so I started to other CI runners.
> 
> The other builds pass. Could you rebase to make sure the apple build is indeed a fluke. The number of errors it gives makes me uncomfortable to merge without a new run.

@mordante I rebased to the current main branch.

mordante commented:
The CI is green. However it seems GitHub wants to merge with a noreply email address, so it seems your address is still hidden.

PaulXiCao commented:
> The CI is green. However it seems GitHub wants to merge with a noreply email address, so it seems your address is still hidden.

I updated my profile and it now has a public email. Hope that helps. Sorry about the issues..

philnik777 commented on libcxx/include/cmath:
```diff
@@ -610,6 +619,161 @@ _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR bool __constexpr_isfinite(_A1 __lcpp_x)
   return __builtin_isfinite(__lcpp_x);
 }
 
+_LIBCPP_CONSTEXPR inline _LIBCPP_HIDE_FROM_ABI float __constexpr_copysign(float __x, float __y) _NOEXCEPT {
```
This code looks like it's been accidentally re-added.

PaulXiCao replied:
Fixed in 6a6fce8a1673355fc72f6ddfbbc1fa3d9d2d9d87.

Zingam commented:
Why does this PR have so few CI checks?

mordante commented:
> > The CI is green. However it seems GitHub wants to merge with a noreply email address, so it seems your address is still hidden.
> 
> @mordante I updated my profile and it now has a public email. Hope that helps. Sorry about the issues..

No problem, I just verified and your e-mail is visible now, thanks.

Can you have a look at @philnik777's last comment.

> Why does this PR have so few CI checks?

I expect since the last failure was apple only.

PaulXiCao committed:
remove wrongly added code

mordante commented:
Thanks a lot again for working on this! I expect it will be faster to get the other functions in.
